### PR TITLE
Add documentation and examples for zap

### PR DESCRIPTION
**Pull request by akshayjshah**
Add a bunch of testable examples and documentation for the main `zap`
package. As we get closer to launching fx and pushing zap adoption
internally, documentation will get more and more important.

We'll still need to improve the documentation of the other packages, but
this addresses the lion's share of #139.

### CODE REVIEW

File: doc.go [hunk 1]
```diff
@@ -18,5 +18,92 @@
...
+//    "attempt", 3,
+//    "backoff", time.Second,
+//  )
+//  sugar.Printf("failed to fetch URL: %s", "http://example.com")
+//
+// In the rare contexts where every microsecond and every allocation matter,
+// use the Logger. It's even faster than the SugaredLogger and allocates far
+// less, but it only supports strongly-typed, structured logging.
+//  logger := zap.NewExample()
+//  defer logger.Sync()
```
Comment (willhug) [inline doc.go]: Can you add an explanation of the logger.Sync somewhere in here?  It's used a lot but I'm not sure when it should be called (i.e. after I grab a logger, after I log a line, etc).

File: example_test.go [hunk 2]
```diff
@@ -0,0 +1,324 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
```
Comment (willhug) [inline example_test.go]: should this be 2017?

File: doc.go [hunk 3]
```diff
@@ -18,5 +18,92 @@
 // OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 // THE SOFTWARE.
 
-// Package zap provides fast, structured, leveled logging in Go.
+// Package zap provides fast, structured, leveled logging.
```
Comment (willhug) [inline doc.go]: do we have a good link that can explain leveled logging?  Wouldn't be surprised if new engineers are not as familiar with the nuances.

File: example_test.go [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+	// zapcore.Cores, then tee the four cores together.
+	core := zapcore.NewTee(
+		zapcore.NewCore(kafkaEncoder, topicErrors, highPriority),
+		zapcore.NewCore(consoleEncoder, consoleErrors, highPriority),
+		zapcore.NewCore(kafkaEncoder, topicDebugging, lowPriority),
+		zapcore.NewCore(consoleEncoder, consoleDebugging, lowPriority),
+	)
+
+	// From a zapcore.Core, it's easy to construct a Logger.
+	logger := zap.New(core)
```
Comment (willhug) [inline example_test.go]: should I always call some form of `zap.New` before I make a log, or should I store a logger on a struct and reuse it.  And if we're in the reuse case, when should I call `logger.Sync`?

File: example_test.go [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+
+	logger.Info("info logging enabled")
+
+	atom.SetLevel(zap.ErrorLevel)
+	logger.Info("info logging disabled")
+	// Output:
+	// {"level":"info","msg":"info logging enabled"}
+}
+
+func ExampleAtomicLevel_config() {
```
Comment (willhug) [inline example_test.go]: should `config` be capitalized?

File: logger.go [hunk 6]
```diff
@@ -203,19 +229,21 @@ func (log *Logger) Panic(msg string, fields ...zapcore.Field) {
...
+// disabled.
 func (log *Logger) Fatal(msg string, fields ...zapcore.Field) {
 	if ce := log.check(FatalLevel, msg); ce != nil {
 		ce.Write(fields...)
 	}
 }
 
-// Sync flushes any buffered log entries.
+// Sync calls the underlying Core's Sync method, flushing any buffered log
+// entries. Applications should take care to call Sync before exiting.
```
Comment (willhug) [inline logger.go]: ahh, so an `OnStop` functionality

File: options.go [hunk 7]
```diff
@@ -22,7 +22,7 @@ package zap
 
 import "go.uber.org/zap/zapcore"
 
-// Option is used to set options for the logger.
+// An Option configures a Logger.
```
Comment (willhug) [inline options.go]: I though gofmt comments needed to start with the name of the interface.

File: example_test.go (hunk 5 above)
Comment (akshayjshah) [inline example_test.go]: Nope.

File: logger.go (hunk 6 above)
Comment (akshayjshah) [inline logger.go]: Yep, will clarify in the top-level examples.

File: options.go (hunk 7 above)
Comment (akshayjshah) [inline options.go]: Nope, preceding articles are allowed.

File: example_test.go (hunk 2 above)
Comment (akshayjshah) [inline example_test.go]: No. It should actually be `2016-2017`, but I'm punting that whole thing until I can write a tool to handle it properly. 

I'll forward you the whole email chain with legal.

File: doc.go (hunk 1 above)
Comment (akshayjshah) [inline doc.go]: Will do.

File: doc.go (hunk 3 above)
Comment (akshayjshah) [inline doc.go]: Errrr...wow, I hadn't thought of that. I don't know of one.

If this comes up in an issue or something, I'll dig up a link and add an FAQ entry.

**Pull request merged**